File: app/src/main/java/soup/movie/ui/theater/sort/TheaterSortActivity.java
package soup.movie.ui.theater.sort;

import android.app.SharedElementCallback;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;
import soup.movie.R;
import soup.movie.ui.BaseActivity;
import soup.movie.ui.theater.edit.TheaterEditActivity;
import soup.widget.drag.ItemTouchHelperAdapter;
import soup.widget.drag.SimpleItemTouchHelperCallback;

import static soup.movie.util.RecyclerViewUtil.verticalLinearLayoutManager;

public class TheaterSortActivity extends BaseActivity<TheaterSortContract.View, TheaterSortContract.Presenter>
        implements TheaterSortContract.View {

    @Inject
    TheaterSortContract.Presenter presenter;

    @BindView(R.id.list)
    RecyclerView listView;

    private TheaterSortListAdapter listAdapter;

    @Override
    protected int getLayoutRes() {
        return R.layout.activity_theater_sort;
    }

    @NonNull
    @Override
    protected TheaterSortContract.Presenter getPresenter() {
        return presenter;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        postponeEnterTransition();
        setEnterSharedElementCallback(new SharedElementCallback() {
            @Override
            public void onMapSharedElements(List<String> names, Map<String, View> sharedElements) {
                for (String name : names) {
                    View child = listView.findViewWithTag(name);
                    sharedElements.put(name, child);
                }
            }
        });
    }

    @Override
    protected void initViewState(@NonNull Context ctx) {
        super.initViewState(ctx);
        listView.setLayoutManager(verticalLinearLayoutManager(this));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.sort_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_edit:
                startActivity(new Intent(this, TheaterEditActivity.class));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void render(@NonNull TheaterSortViewState viewState) {
        ItemTouchHelperAdapter adapterDelegate = new ItemTouchHelperAdapter() {
            @Override
            public void onItemMove(int fromPosition, int toPosition) {
                listAdapter.onItemMove(fromPosition, toPosition);
            }

            @Override
            public void onItemDismiss(int position) {
                listAdapter.onItemDismiss(position);
            }
        };
        ItemTouchHelper.Callback callback = new SimpleItemTouchHelperCallback(adapterDelegate);
        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(callback);
        itemTouchHelper.attachToRecyclerView(listView);

        listAdapter = new TheaterSortListAdapter(
                viewState.getSelectedTheaters(), itemTouchHelper::startDrag);
        listView.setAdapter(listAdapter);
        startPostponedEnterTransition();
    }

    @OnClick(R.id.button_cancel)
    public void onCancelClicked() {
        onBackPressed();
    }

    @OnClick(R.id.button_confirm)
    public void onConfirmClicked() {
        presenter.onConfirmClicked(listAdapter.getSelectedTheaters());
        onBackPressed();
    }
}

File: app/src/main/java/soup/movie/ui/theater/sort/TheaterSortActivity.kt
package soup.movie.ui.theater.sort

import android.app.ActivityOptions
import android.content.Intent
import android.os.Bundle
import android.util.Pair
import android.view.View
import androidx.core.app.SharedElementCallback
import androidx.core.view.isVisible
import androidx.core.view.postOnAnimationDelayed
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.listener.OnDragStartListener
import androidx.recyclerview.widget.listener.OnItemMoveListener
import androidx.recyclerview.widget.util.SimpleItemTouchHelperCallback
import com.google.android.material.chip.Chip
import kotlinx.android.synthetic.main.activity_theater_sort.*
import soup.movie.R
import soup.movie.databinding.ActivityTheaterSortBinding
import soup.movie.ui.BaseActivity
import soup.movie.ui.theater.edit.TheaterEditActivity
import soup.movie.util.observe
import soup.movie.util.with

class TheaterSortActivity : BaseActivity() {

    private val viewModel: TheaterSortViewModel by viewModel()

    private val listAdapter: TheaterSortListAdapter by lazy {
        val callback = SimpleItemTouchHelperCallback(object : OnItemMoveListener {
            override fun onItemMove(fromPosition: Int, toPosition: Int) {
                listAdapter.onItemMove(fromPosition, toPosition)
                viewModel.onItemMove(fromPosition, toPosition)
            }
        })
        val itemTouchHelper = ItemTouchHelper(callback).apply {
            attachToRecyclerView(listView)
        }
        TheaterSortListAdapter(object : OnDragStartListener {
            override fun onDragStart(viewHolder: RecyclerView.ViewHolder) {
                itemTouchHelper.startDrag(viewHolder)
            }
        })
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        DataBindingUtil.setContentView<ActivityTheaterSortBinding>(this, R.layout.activity_theater_sort).apply {
            lifecycleOwner = this@TheaterSortActivity
        }

        listView.adapter = listAdapter

        //FixMe: find a timing to call startPostponedEnterTransition()
        listView.postOnAnimationDelayed(100) {
            startPostponedEnterTransition()
        }

        postponeEnterTransition()
        setEnterSharedElementCallback(object : SharedElementCallback() {
            override fun onMapSharedElements(names: List<String>,
                                             sharedElements: MutableMap<String, View>) {
                sharedElements.clear()
                listView?.run {
                    (0 until childCount)
                        .mapNotNull { getChildAt(it) }
                        .mapNotNull { it.findViewById<Chip>(R.id.theaterChip) }
                        .forEach { sharedElements[it.transitionName] = it }
                }
            }
        })
        setExitSharedElementCallback(object : SharedElementCallback() {
            override fun onMapSharedElements(names: List<String>,
                                             sharedElements: MutableMap<String, View>) {
                sharedElements.clear()
                names.forEach { name ->
                    listView.findViewWithTag<View>(name)?.run {
                        sharedElements[name] = this
                    }
                }
            }
        })

        viewModel.uiModel.observe(this) {
            render(it)
        }
    }

    override fun onActivityReenter(resultCode: Int, data: Intent?) {
        super.onActivityReenter(resultCode, data)
        postponeEnterTransition()

        //FixMe: find a timing to call startPostponedEnterTransition()
        listView.postOnAnimationDelayed(100) {
            startPostponedEnterTransition()
        }
    }

    private fun render(uiModel: TheaterSortUiModel) {
        listAdapter.submitList(uiModel.selectedTheaters)
        noItemsView.isVisible = uiModel.selectedTheaters.isEmpty()
    }

    override fun onBackPressed() {
        viewModel.saveSnapshot()
        setResult(RESULT_OK)
        super.onBackPressed()
    }

    fun onAddItemClick(view: View) {
        val intent = Intent(this, TheaterEditActivity::class.java)
        startActivityForResult(intent, 0, ActivityOptions
            .makeSceneTransitionAnimation(this, *createSharedElements())
            .toBundle())
    }

    private fun createSharedElements(): Array<Pair<View, String>> =
        listView?.run {
            (0 until childCount)
                .mapNotNull { getChildAt(it) }
                .mapNotNull { it.findViewById<Chip>(R.id.theaterChip) }
                .map { it with it.transitionName }
                .toTypedArray()
        } ?: emptyArray()
}
